### PR TITLE
Support cursors in FakeFileSystem sources

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Review by oldergod — APPROVED** [review]

File: okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt [hunk 1]
```diff
@@ -437,17 +438,51 @@ class FakeFileSystem(
...
+      return size - buffer.size
+    }
+
+    override fun size(): Long {
+      check(!closed) { "closed" }
+      return size
+    }
+
+    override fun seek(position: Long) {
+      check(!closed) { "closed" }
```
Comment (yschimke) [inline okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt:471]: The thing I love here about FakeFileSystem is that the default behaviour is the most strict (allowReadsWhileWriting) which will flush out issues in testing, while being more lenient in production.  

Reviewing this code, I'm less certain whether I should worry about thread concurrency.

Should by default a FakeFileSource be locked to a single thread by default, simplifying reviewing? Potentially throwing if threads swap, or if we detect concurrency?

**Comment by yschimke** [comment]
Any idea why it's failing on 2x JDK 8 CI builds?

```
okio.SystemFileSystemTest[jvm] > sourceCursorWhenClosed[jvm] FAILED
    java.lang.AssertionError at Assert.java:88
```

File: okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt (hunk 1 above)
Comment (swankjesse) [reply okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt:471]: Unfortunately that'll break producer consumer models where one thread picks files to read and another thread reads em.

But we could add an opt-in for callers that know they don't do that?

strict concurrency()

**Comment by swankjesse** [comment]
I think the exception we're checking for is different on JDK 8.

File: okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt (hunk 1 above)
Comment (yschimke) [reply okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt:471]: My difficulty of the review is how to review whether this code is concurrency safe. Maybe on first read it makes the thread sticky is strictConcurrency is set?  Also in that case which thread creates the source instances?

Comment (swankjesse) [reply okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt:471]: ZipFileSystem is intended to be immutable once created. Individual sources are not safe for concurrent use.

**Review by yschimke — APPROVED** [review]

**Pull request merged**